drv/opentitan-aontimer/src/lib.rs

```diff
@@ -0,0 +1,150 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#![no_std]
+
+#[allow(unused_imports)]
+use userlib::*;
+// use zerocopy::AsBytes;
+
+use core::ptr::slice_from_raw_parts_mut;
+
+const WDOG_REGWEN: usize = 4;
+const WDOG_CTRL: usize = 5;
+const WDOG_BARK_THOLD: usize = 6;
+const WDOG_BITE_THOLD: usize = 7;
+const WDOG_COUNT: usize = 8;
+const INTR_STATE: usize = 9;
+
+// The OpenTitan Always-On Timer is a counter that increments. If the counter
+// reaches a configurable threshold, it will emit an interrupt signal, called the
+// "bark". If the counter reaches a second configurable threshold, it will reset
+// the system, called the "bite".
+// Enabling the AON Timer before setting a bite threshold causes the system to
+// immediately reset.
+pub struct AonTimer {
+    base_addr: *mut [u32],
+    clock_freq: u32,
+    pub bark_cb: Option<fn()>,
+}
+
+impl AonTimer {
+    // Initialize a new watchdog driver device with the specified base address,
+    // clock frequency (in Hz), and the bark/bite thresholds (both in ms).
+    // If the bark threshold > bite threshold, the bite threshold is set to
+    // the bark threshold.
+    pub fn new(
+        base_addr: u32,
+        clock_freq: u32,
+        bark_thold_ms: u64,
+        bite_thold_ms: u64,
+        bark_cb: Option<fn()>,
+    ) -> Self {
+        let inst = AonTimer {
+            base_addr: slice_from_raw_parts_mut(base_addr as *mut u32, 12),
+            clock_freq,
+            bark_cb,
+        };
+        unsafe {
+            //1: init regs
+            (*inst.base_addr)[WDOG_COUNT] = 0;
+            (*inst.base_addr)[INTR_STATE] = 0b11;
+            (*inst.base_addr)[WDOG_BARK_THOLD] = inst.ms_to_reg(bark_thold_ms);
+            if bark_thold_ms > bite_thold_ms {
+                (*inst.base_addr)[WDOG_BITE_THOLD] =
+                    inst.ms_to_reg(bark_thold_ms);
+            } else {
+                (*inst.base_addr)[WDOG_BITE_THOLD] =
+                    inst.ms_to_reg(bite_thold_ms);
```

Comment:
Should we impose this on the user without it being clear that `bite_thold_ms` must be greater than `bark_thold_ms`?

Reply:
Added compile time check